src/lib.rs

```diff
@@ -25,40 +27,31 @@ pub use term::load_book::load_file_to_book;
 pub const ENTRY_POINT: &str = "main";
 pub const HVM1_ENTRY_POINT: &str = "Main";
 
-pub fn check_book(mut book: Book) -> Result<(), String> {
+pub fn check_book(book: Book) -> Result<(), String> {
   // TODO: Do the checks without having to do full compilation
-  compile_book(&mut book, CompileOpts::light(), None)?;
+  compile_book(book, CompileOpts::light())?;
   Ok(())
 }
 
-pub fn compile_book(
-  book: &mut Book,
-  opts: CompileOpts,
-  entrypoint: Option<Name>,
-) -> Result<CompileResult, String> {
-  let (main, warnings) = desugar_book(book, opts, entrypoint)?;
-  let (nets, hvmc_names, labels) = book_to_nets(book, &main);
+pub fn compile_book(mut book: Book, opts: CompileOpts) -> Result<CompileResult, String> {
+  let main = desugar_book(&mut book, opts)?;
+  let (nets, hvmc_names, labels) = book_to_nets(&mut book, &main);
   let mut core_book = nets_to_hvmc(nets, &hvmc_names)?;
   if opts.pre_reduce {
     pre_reduce_book(&mut core_book, opts.pre_reduce_refs, book.hvmc_entrypoint())?;
   }
   if opts.prune {
     prune_defs(&mut core_book, book.hvmc_entrypoint());
   }
-  Ok(CompileResult { core_book, hvmc_names, labels, warnings })
+  Ok(CompileResult { book, core_book, hvmc_names, labels })
 }
 
-pub fn desugar_book(
-  book: &mut Book,
-  opts: CompileOpts,
-  entrypoint: Option<Name>,
-) -> Result<(Name, Vec<Warning>), String> {
-  let mut warnings = Vec::new();
-  let main = book.check_has_entrypoint(entrypoint)?;
-  book.check_shared_names()?;
+pub fn desugar_book(book: &mut Book, opts: CompileOpts) -> Result<Name, String> {
```

Comment:
Wouldn't it make more sense to return the errors here and call the stringification in the CLI?
Since we're already introducing the error types we might as well expose them for whoever wants to use them

Reply:
Or rather the set of diagnostics `Info`